src/Features/LanguageServer/Protocol/Handler/Formatting/AbstractFormatDocumentHandlerBase.cs

```diff
@@ -25,13 +27,22 @@ internal abstract class AbstractFormatDocumentHandlerBase<RequestType, ResponseT
             RequestContext context,
             LSP.FormattingOptions options,
             IGlobalOptionService globalOptions,
-            CancellationToken cancellationToken,
-            LSP.Range? range = null)
+            LSP.Range? range,
+            CancellationToken cancellationToken)
         {
-            var document = context.Document;
-            if (document is null)
+            if (context.Document is not { } document)
                 return null;
 
+            IList<TextChange>? textChanges = null;
+            if (range is null && globalOptions.GetOption(LspOptionsStorage.LspFormattingSortImports, document.Project.Language))
+            {
+                var organizeImports = document.GetRequiredLanguageService<IOrganizeImportsService>();
+                var organizeImportsOptions = await document.GetOrganizeImportsOptionsAsync(globalOptions, cancellationToken).ConfigureAwait(false);
+                var organizedDocument = await organizeImports.OrganizeImportsAsync(document, organizeImportsOptions, cancellationToken).ConfigureAwait(false);
```

Comment:
If OrganizeImportsAsync makes no changes to the document, does it return the same version that was passed in? If that is the case, can we delay getting text changes until the very end and only call it once.

Reply:
I believe we do need to get only the final changes - the spec mentions that edits returned as a `TextEdit[]` should not overlap - https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEditArray

If we get them separately it seems possible that a formatting change overlaps an organize import change?